src/go/sysfail.go

```diff
@@ -0,0 +1,186 @@
+package sysfail
+
+/*
+#cgo CFLAGS: -I../../include
+#cgo LDFLAGS: -L. -lsysfail
+
+#include "sysfail.h"
+#include <stdlib.h>
+
+void set_poll_itvl(sysfail_thread_discovery_t* disc, uint32_t value) {
+    disc->poll_itvl_usec = value;
+}
+
+void stop_sysfail(sysfail_session_t* session) {
+    if (session) session->stop(session);
+}
+
+void add_this_thread(sysfail_session_t* session) {
+    if (session) session->add_this_thread(session);
+}
+
+void remove_this_thread(sysfail_session_t* session) {
+    if (session) session->remove_this_thread(session);
+}
+
+void add_thread(sysfail_session_t* session, sysfail_tid_t tid) {
+    if (session) session->add_thread(session, tid);
+}
+
+void remove_thread(sysfail_session_t* session, sysfail_tid_t tid) {
+    if (session) session->remove_thread(session, tid);
+}
+
+void discover_threads(sysfail_session_t* session) {
+    if (session) session->discover_threads(session);
+}
+
+void set_error_weights(sysfail_outcome_t* outcome, sysfail_error_wt_t* weights) {
+	outcome->error_wts = weights;
+}
+
+*/
+import "C"
+import (
+	"unsafe"
+)
+
+type Probability struct {
+	P         float64
+	AfterBias float64
+}
+
+type ErrorWeight struct {
+	Nerror int
+	Weight float64
+}
+
+type Outcome struct {
+	Fail         Probability
+	Delay        Probability
+	MaxDelayUsec uint32
+	Ctx          unsafe.Pointer
+	Eligible     func(ctx unsafe.Pointer, regs *C.greg_t) int
+	NumErrors    uint32
+	ErrorWeights []ErrorWeight
+}
+
+type ThreadDiscoveryStrategy uint32
+
+const (
+	ThreadDiscoveryNone ThreadDiscoveryStrategy = iota
+	ThreadDiscoveryPoll
+)
+
+type ThreadDiscoveryConfig struct {
+	PollIntervalUsec uint32
+}
+
+type SyscallOutcome struct {
+	Syscall int
+	Outcome Outcome
+}
+
+type Plan struct {
+	Strategy        ThreadDiscoveryStrategy
+	Config          ThreadDiscoveryConfig
+	Ctx             unsafe.Pointer
+	Selector        func(ctx unsafe.Pointer, tid C.sysfail_tid_t) int
+	SyscallOutcomes []*SyscallOutcome
+}
+
+type Session struct {
+	session *C.sysfail_session_t
+	plan    *C.sysfail_plan_t
+}
+
+// NewPlan creates a new plan for failure injection.
+func NewPlan(strategy ThreadDiscoveryStrategy, pollIntervalUsec uint32, syscallOutcomes []*SyscallOutcome) *Plan {
+	return &Plan{
+		Strategy: strategy,
+		Config: ThreadDiscoveryConfig{
+			PollIntervalUsec: pollIntervalUsec,
+		},
+		SyscallOutcomes: syscallOutcomes,
+		Selector:        nil, // Use appropriate selector function
+		Ctx:             nil, // User data
+	}
+}
+
+// StartSession starts a new failure injection session.
+func StartSession(plan *Plan) *Session {
+	var cOutcomes *C.sysfail_syscall_outcome_t
```

Comment:
Do we need to initialize to nil or 0 for the no-outcomes case? May be golang does on its own, reviewing on phone, so can't check.

Reply:
I think it does it by default, but initialized it to nil just in case.